e3nn_jax/_src/linear_equinox.py

```diff
@@ -0,0 +1,347 @@
+        num_indexed_weights (optional int): number of indexed weights. See example below.
+        weights_per_channel (bool): whether to have one set of weights per channel.
+        force_irreps_out (bool): whether to force the output irreps to be the one specified in ``irreps_out``.
+
+    Due to how Equinox is implemented, the random key, irreps_in and irreps_out must be supplied at initialization.
+    The type of the linear layer must also be supplied at initialization:
+    'vanilla', 'indexed', 'mixed', 'mixed_per_channel'
+    Also, depending on what type of linear layer is used, additional options
+    (eg. 'num_indexed_weights', 'weights_per_channel', 'weights_dim', 'channel_in')
+    must be supplied.
+
+    Examples:
+        Vanilla::
+
+            >>> import e3nn_jax as e3nn
+            >>> import jax
+
+            >>> x = e3nn.normal("0e + 1o")
+            >>> linear = e3nn.equinox.Linear(
+                    irreps_out="2x0e + 1o + 2e",
+                    irreps_in=x.irreps,
+                    key=jax.random.PRNGKey(0),
+                )
+            >>> linear(x).irreps  # Note that the 2e is discarded. Avoid this by setting force_irreps_out=True.
+            2x0e+1x1o
+            >>> linear(x).shape
+            (5,)
+
+        External weights::
+
+            >>> linear = e3nn.equinox.Linear(
+                    irreps_out="2x0e + 1o",
+                    irreps_in=x.irreps,
+                    linear_type="mixed",
+                    weights_dim=4,
+                    key=jax.random.PRNGKey(0),
+                )
+            >>> e = jnp.array([1., 2., 3., 4.])
+            >>> linear(e, x).irreps
+                2x0e+1x1o
+            >>> linear(e, x).shape
+            (5,)
+
+        Indexed weights::
+
+            >>> linear = e3nn.equinox.Linear(
+                    irreps_out="2x0e + 1o + 2e",
+                    irreps_in=x.irreps,
+                    linear_type="indexed",
+                    num_indexed_weights=3,
+                    key=jax.random.PRNGKey(0),
+                )
+            >>> i = jnp.array(2)
+            >>> linear(i, x).irreps
+                2x0e+1x1o
+            >>> linear(i, x).shape
+            (5,)
+    """
+    irreps_out: e3nn.Irreps
+    irreps_in: e3nn.Irreps
+    channel_out: int
+    channel_in: int
+    gradient_normalization: Optional[Union[float, str]]
+    path_normalization: Optional[Union[float, str]]
+    biases: bool
+    num_indexed_weights: Optional[int]
+    weights_per_channel: bool
+    force_irreps_out: bool
+    weights_dim: Optional[int]
+    linear_type: str
+
+    # These are used internally.
+    _linear: FunctionalLinear
+    _weights: Dict[str, jnp.ndarray]
+    _output_irreps_unsimplified: e3nn.Irreps
+    _output_irreps: e3nn.Irreps
+    _input_dtype: jnp.dtype
+
+    def __init__(
+        self,
+        irreps_out: e3nn.Irreps,
+        irreps_in: e3nn.Irreps,
```

Comment:
What about in before out?

Reply:
The Flax and Haiku ones have out before in, so I thought it would be better to be consistent

Reply:
Flax and Haiku don't require `irreps_in` to be specified. Actually we could even remove `irreps_in` in the Flax and Haiku

Reply:
But also the convention would change `e3nn.equinox.Linear("1o")` would mean something different from `e3nn.flax.Linear("1o")`. I would prefer to keep it as is.

Reply:
`e3nn.equinox.Linear("1o")` would crash right?

But as you wish, I can merge like that